File: quiz/deno/src/statemaschine/states/quiz/participate/join-quiz-state.ts
import { IState } from '@states/state.ts';
import { Input } from '@cliffy/prompt/mod.ts';
import { Quiz } from '@src/quiz/quiz.ts';
import { User } from '@src/user.ts';
import { keypress } from '@cliffy/keypress/mod.ts';
import { RunQuizState } from '@states/quiz/participate/run-quiz-state.ts';
import { UI } from '@src/utils.ts';

export class JoinQuizState implements IState {
	async run(): Promise<IState> {
		const availableShareCodes = await Quiz.availableShareCodes();

		const shareCode = await Input.prompt({
			message: 'Enter the share code:',
			validate: (shareCode) => {
				if (!availableShareCodes.includes(shareCode)) {
					return 'Invalid share code';
				}
				return true;
			},
		});

		const quiz = await Quiz.getQuizByShareCode(shareCode);
		const username = await User.getUsernameById(quiz?.userId ?? '');

		if (!quiz || !username) {
			throw new Error('Invalid quiz or user');
		}

		UI.clear();

		UI.prompt(
			`%cWelcome to '${quiz.name}' by '${username}`,
			'color: #00f; font-weight: bold;',
		);
		if (quiz.description) {
			UI.prompt(quiz.description);
		}
		UI.prompt();
		UI.prompt('Press any key to start the quiz');
		await keypress();

		return new RunQuizState(quiz);
	}
}

File: quiz/go/statemaschine/states/join-quiz-state.go
package states

import (
	"errors"

	"github.com/DoktorLenz/kp-wise23-24/quiz/go/quiz"
	"github.com/DoktorLenz/kp-wise23-24/quiz/go/user"
	"github.com/DoktorLenz/kp-wise23-24/quiz/go/utils"
	"github.com/manifoldco/promptui"
)

type JoinQuizState struct {
	IState
}

func (state *JoinQuizState) Run() IState {

	availableShareCodes, err := quiz.AvailableShareCodes()
	if err != nil {
		panic(err)
	}

	prompt := promptui.Prompt{
		Label: "Enter the share code:",
		Validate: func(shareCode string) error {
			for _, code := range availableShareCodes {
				if code == shareCode {
					return nil
				}
			}
			return errors.New("Invalid share code")
		},
	}

	shareCode, _ := prompt.Run()

	quiz, err := quiz.GetQuizByShareCode(shareCode)
	if err != nil {
		panic(err)
	}

	username, err := user.GetUsernameByID(quiz.UserID)
	if err != nil {
		panic(err)
	}

	utils.Clear()
	utils.Prompt("Welcome to " + quiz.Name + " by " + username + "!")
	if quiz.Description != "" {
		utils.Prompt(quiz.Description)
	}
	utils.Prompt("")
	utils.Prompt("Press enter to start the quiz...")
	utils.AwaitInput()

	return &RunQuizState{quiz: quiz}
}
